### ISSUE TITLE
One way audio on SRTP calls after ~35 minutes. Discarded invalid SRTP packet: authentication failed

### ISSUE THREAD

**Issue by arkadiam** [sipwise/rtpengine#917]
Test environment:
Ubuntu 16.04.6 LTS, Kamailio + Rtpengine, Endpoints: Acrobits softphone, Linphone softphone, Polycom phone.
Tested rtpengine versions:
Version: 6.3.0.0+0~mr6.3.0.0 git-master-d4763ab, master HEAD.

User experience:
Acrobits softphone makes a call. In about 35 min after the call is answered the called party stops hearing calling party. At this moment rtpengine logs: "Discarded invalid SRTP packet: authentication failed". It looks like an audio sent by a softphone can't be decrypted by the rtpengine after 35 min mark and being discarded.

Rtpengine started as:
```
/usr/sbin/rtpengine -f -E --no-log-timestamps --pidfile /run/ngcp-rtpengine-daemon.pid --config-file /etc/rtpengine/rtpengine.conf --table 0
```
Config `/etc/rtpengine/rtpengine.conf` has:
`table = 0`

In the case when Acrobits endpoint is used, rtpengine does OPUS-PCMU codec conversion and SRTP-RTP conversions.

When testing other SIP endpoints there is no OPUS-PCMU conversion, only SRTP-RTP and there is no audio interruption at 35 min mark.  There is also no audio interruption for any endpoints when rtpengine started with `table = -1`.

```
```

I've introduces logging at `daemon/call.c call_timer()`, and `daemon/rtp.c rtp_savp2avp()`, and `packet_index()` to observe how SRTP index and other parameters are calculated.

There is this code in `daemon/call.c call_timer()`:
```
        if (sfd->crypto.params.crypto_suite && ps->ssrc_in
                        && ntohl(ke->target.ssrc) == ps->ssrc_in->parent->h.ssrc
                        && ke->target.decrypt.last_index - ps->ssrc_in->srtp_index > 0x4000)
        {
                ps->ssrc_in->srtp_index = ke->target.decrypt.last_index;
                update = 1;
        }
```
For the scenario when there is opus-pcmu + srtp-rtp conversion, every second, this code resets `ps->ssrc_in->srtp_index` to a value that was obtained at the begging of the call.

According to my trace `ke->target.decrypt.last_index` value is never changed for the call. 
So srtp_index is always reset to some low value.  After a wile, when the true SRTP index expected to be large, `srtp_index` reset, starts confusing index calculation in `rtp_savp2avp()` and `packet_index()`.
Very low value of `srtp_index` makes `v` and `roc` equal to `0` when they supposed to be `2` or greater.

Current `rtp_savp2avp()` code, is able to recover SRTP index while user is up to ~35 minutes in a call. 
After that time index is not recovering correctly. 

I.e. `rtp_savp2avp()` index 'recovery' technique:
```
    index += 0x10000;
```
works for up to 35 min. 

Past that time, `c->params.crypto_suite->hash_rtp()` validation fails and packet discarded as invalid.

As a "hack", to extend the two way audio past 35 minutes, we can apply `index += 0x10000;` multiple times. 
Depending on the duration of the call, index validation will succeed with one value or another. This is a guess work and extra computation work per packet when the call is long.

It seem to me that the reason why `rtp_savp2avp()` applies this `+0x10000` correction is to counterbalance the code in `daemon/call.c call_timer()`:
```
        if (sfd->crypto.params.crypto_suite && ps->ssrc_in
                        && ntohl(ke->target.ssrc) == ps->ssrc_in->parent->h.ssrc
                        && ke->target.decrypt.last_index - ps->ssrc_in->srtp_index > 0x4000)
        {
                ps->ssrc_in->srtp_index = ke->target.decrypt.last_index;
                update = 1;
        }
```
and it works only to a certain point.

When the call is properly kernelized or if there is no kenelization happening this 'reset' never happens:
```
        ps->ssrc_in->srtp_index = ke->target.decrypt.last_index;
```
and `rtp_savp2avp()` and `packet_index()` are able to keep up with SRTP index tracking.

Can you explain when  `ps->ssrc_in->srtp_index = ke->target.decrypt.last_index;` is expected to be called? Is it possible to remove this assignment or to make `if()` condition more specific?
I've tested a build with this line commented, and there in no media interruption with Acrobits client anymore.

For Acrobits softphone, with opus-pcmu+srtp-rtp conversion, rtpengine says that media streams are kernelized but rtp_savp2avp() is still being called for every packet. I haven't noticed any error that would say that kernelization wasn't successful.
How does codec transcoding affect the decision to handle stream in kernel vs userspace?

```
```
Summary:
- one way audio after 35 min with Acrobits endpoint
- rtpengine does opus-pcmu and srtp-rtp conversion
- rtpengine started with kernel module support and table 0.
- call is not fully kernelized (probably prevented by opus transcoding)
- `ps->ssrc_in->srtp_index = ke->target.decrypt.last_index;`  in `call_timer()` being called every second with the same value, which messes up `rtp_savp2avp()` algorithm after SRTP packet index is expected to be greater then `131072`.
- no issue when stream is handled by kernel module only or in userspace only.
- possibly related to #209, #71, #674

**Comment by rfuchs** [comment]
Thanks for the detailed report. I think I already know what's going on. The RTP stream ends up partially kernelised, for only certain payload types. Other ones are left delegated to userspace for transcoding. However, the loop updating the SRTP index from the kernel doesn't know about this distinction and expects to have the SRTP index updated from the kernel for all payload types, which would overwrite the userspace tracking of the SRTP index. I'll prepare a fix.

**Comment by arkadiam** [comment]
Thanks for a quick reply. Looking forward to test your patch.

**Comment by rfuchs** [comment]
See patch above. Completely untested.

**Comment by arkadiam** [comment]
I've applied the patch and running rtpengine with kernelspace support. Now `ps->ssrc_in->srtp_index =...` is not being called, srtp index is not reset, and there was two way audio on 1h+ long call. So it looks like this patch solves the issue. Thanks!

I've noticed `rtp_savp2avp()` is being called for every packet. The stream is partially kernelized. Is it the expected behaviour? I have opus-pcmu + srtp-rtp transcoding on this call.
